sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionals.scala

```diff
@@ -296,9 +295,7 @@ case class CaseKeyWhen(key: Expression, branches: Seq[Expression]) extends CaseW
   }
 
   private def equalNullSafe(l: Any, r: Any) = {
-    if (l == null && r == null) {
-      true
-    } else if (l == null || r == null) {
+    if (l == null || r == null) {
```

Comment:
This is far enough from Java/Scala semantics that it warrants a scaladoc and a pointer to NULL semantics in SQL, IMO